pkg/controller/storagecluster/placement.go

```diff
@@ -20,16 +20,20 @@ func getPlacement(sc *ocsv1.StorageCluster, component string) rookv1.Placement {
 		in := defaults.DaemonPlacements[component]
 		(&in).DeepCopyInto(&placement)
 	}
-	if sc.Spec.LabelSelector == nil {
+
+	// If no placement is specified for the given component and the
+	// StorageCluster has no label selector, set the default node
+	// affinity.
+	if placement.NodeAffinity == nil && sc.Spec.LabelSelector == nil {
 		placement.NodeAffinity = defaults.DefaultNodeAffinity
-	} else {
-		term := convertLabelToNodeSelector(*sc.Spec.LabelSelector)
-		if len(term.MatchExpressions) != 0 {
-			placement.NodeAffinity = &corev1.NodeAffinity{
-				RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-					NodeSelectorTerms: []corev1.NodeSelectorTerm{term},
-				},
-			}
+	}
+
+	// If the StorageCluster specifies a label selector, append it to the
+	// node affinity, creating it if it doesn't exist.
+	if sc.Spec.LabelSelector != nil {
+		reqs := convertLabelToNodeSelectorRequirements(*sc.Spec.LabelSelector)
+		if len(reqs) != 0 {
+			appendNodeRequirements(&placement, reqs...)
```

Comment:
I am still somewhat confused that new `appendNodeRequirements()` is invoked here in `getPlacement()` but also in `newStorageClassDeviceSets()`, after the invocation of `getPlacement()`. Is this redundant? :thinking: ...
At least the code is still too convoluted to easily understand all the cases.

Reply:
A StorageClassDeviceSet can take a TopologyKey value that must alter the generated Placement for all OSDs to be specific to the device set. As such, it has an additional thing to append after the initial Placement generation. To condense this further I'd have to modify `getPlacement()` to take an optional TopologyKey parameter, which was more refactoring than I thought was necessary.